### DISCUSSION TITLE
this.actions.die() kills multiple actors.

### DISCUSSION THREAD

**Discussion by nparoski** [excaliburjs/Excalibur#2743] [General]
I have player actor which spawns projectile actors which have range actors, when Projectile stops colliding with Range they both need to die. Problem I have is all of projectiles die. 

Projectile 1. and Range1. (blue circle) need to die, but both do: image e.g
![image](https://github.com/excaliburjs/Excalibur/assets/35550355/aaaac268-5cdf-41b5-86f2-574448cddbd8)

**Simplified code example:**
_Note: must click outside circle_
```

import { Color, Actor, Engine, vec, Vector } from 'excalibur';

const game = new Engine({
  width: 800,
  height: 600,
  backgroundColor: Color.Black,
});

class Player extends Actor {
  constructor() {
    super({
      width: 38,
      height: 64,
      color: Color.Cyan,
      pos: vec(125, 125),
    });
  }

  _initProjectile(_engine: Engine, moveTo: Vector) {
    const projectile = new Actor({
      width: 24,
      height: 24,
      color: Color.Red,
      pos: this.pos,
    });

    const range = new Actor({
      radius: 200,
      color: Color.fromRGB(25, 255, 25, 0.5),
      pos: this.pos,
      z: -1,
    });

    projectile.actions.moveTo(moveTo, 200);

    range.on('collisionend', () => {
      projectile.actions.clearActions();
      projectile.actions.die();
      range.actions.die();
    });

    _engine.add(projectile);
    _engine.add(range);
  }

  onInitialize(_engine: Engine): void {
    _engine.input.pointers.on('down', (evt) => {
      if (evt.button == 'Left') {
        this._initProjectile(_engine, evt.worldPos);
      }
    });
  }
}

game.add(new Player());

game.start();

```

**Comment by eonarheim** [comment]
Hi @nparoski, good question this is a tricky one!

I think the issue is that the `range.on('collisionend', ...)` will fire on _any_ collision end event. So for example you have 2 projectiles, when the first triggers `collisionend` both handlers for each projectile will fire.

I think one way to work around this is to invert the logic and have `projectile.on('collisionend', ...)` that way the handler will fire per projectile.

Another way might be to check the `evt` arg to verify it's the projectile you want to kill

```typescript
range.on('collisionend', (evt) => { 
  if (evt.other ===  projectile) {
    // kill
  }

})
```

**Reply by nparoski** [reply]
Makes sense, but I would expect this event to be fired when that specific actor collisionended rather than all actors on screen 🤔

**Reply by eonarheim** [reply]
> Makes sense, but I would expect this event to be fired when that specific actor collisionended rather than all actors on screen 🤔

It's a little more subtle, since `range` is experiencing a `collisionend` event from some actor, without using `evt.other` the handler doesn't know which closed over projectile is the one that "ended" it's collision.

Another strategy you can use if your range is always circular you could add a `projectile.onPostUpdate = () => {...}` hook to check the distance between the player and the projectile.

```typescript
projectile.onPostUpdate = () => {
   const distance = projectile.pos.distance(player.pos);
   if (distance > 200) {
      projectile.kill();
   }
}
```

**Reply by nparoski** [reply]
@eonarheim I'm actually trying to make projectile always collide with range. 

When mouse is inside of circle projectile should go beyond and die when colliding with range.

Struggling with math little bit but I should be able to calculate it.
